### ISSUE TITLE
Unable to add local tracks to local playlist

### ISSUE THREAD

**Issue by jplegat** [jaedb/Iris#657]
**Unable to add local tracks to local playlist**

**To Reproduce**

**Make sure to create a new local playlist before trying to add a local track to the playlist**

1. Enable Test Mode (_Settings > Debug > Test Mode_).
2. Go to 'Tracks'
3. Select any track and click on '...' next to the track name.
4. Select ' Add to Playlist' and add the track to the selected playlist.
5. Track is not added to the playlist - no error or confirmation message is returned.

**Expected behavior**
Track is added to the local playlist

**Console log**
```
Test mode enabled, using un-minified code debug:134:11
Download the React DevTools for a better development experience: https://fb.me/react-devtools react-dom.development.js:24994
redux-persist/stateReconciler: rehydrated keys 'mopidy, lastfm, genius, google, snapcast, _persist' autoMergeLevel1.js:35:43
redux-persist/stateReconciler: rehydrated keys 'outputs, queue, queue_metadata, current_track, pinned, http_streaming_enabled, http_streaming_cachebuster, http_streaming_url, seeds_resolved, radio, current_tracklist, current_track_uri, stream_title, _persist' autoMergeLevel1.js:35:43
redux-persist/stateReconciler: rehydrated keys 'language, theme, smooth_scrolling_enabled, hotkeys_enabled, playback_controls_touch_enabled, allow_reporting, wide_scrollbar_enabled, slim_mode, suppressed_broadcasts, grid_glow_enabled, window_title, uri_schemes_search_enabled, initial_setup_complete, _persist' autoMergeLevel1.js:35:43
redux-persist/stateReconciler: rehydrated keys 'host, port, ssl, current_server, servers, mute, volume, play_state, uri_schemes, library_albums_uri, library_artists_uri, consume, random, repeat, _persist' autoMergeLevel1.js:35:43
redux-persist/stateReconciler: rehydrated keys 'country, locale, enabled, _persist' autoMergeLevel1.js:35:43
redux-persist/stateReconciler: rehydrated keys 'connected, username, client_id, version, config, connecting, commands, pinned, upgrading, _persist, connection_id' autoMergeLevel1.js:35:43
Restoring "mopidy:library:albums" and 3 items from database middleware.js:545:20
Loading more: 3 LazyLoadListener.js:51:20
"local:album:md5:7af582e4cd40f18359c0857842682c15" already in index library.js:104:14
[react-ga] ReactGA.initialize must be called first or GoogleAnalytics should be loaded manually warn.js:2
Restoring "mopidy:library:playlists" and 1 items from database middleware.js:545:20
Uncaught (in promise) 
{…}
​
closeEvent: close { target: WebSocket, isTrusted: true, wasClean: true, … }
​
message: "WebSocket closed"
​
name: "ConnectionError"
​
stack: ""
​
<prototype>: Object { constructor: ConnectionError(message), stack: "" }

Restoring "mopidy:library:tracks" and 29 items from database middleware.js:545:20
mopidy:library:albums" already in index middleware.js:538:18
Loading more: 3 LazyLoadListener.js:51:20
"local:album:md5:7af582e4cd40f18359c0857842682c15" already in index library.js:104:14
mopidy:library:tracks" already in index middleware.js:538:18
[react-ga] ReactGA.initialize must be called first or GoogleAnalytics should be loaded manually warn.js:2

Uncaught (in promise) 
{…}
​
closeEvent: close { target: WebSocket, isTrusted: true, wasClean: true, … }
​
message: "WebSocket closed"
​
name: "ConnectionError"
​
stack: ""
​
<prototype>: {…}
​​
constructor: ConnectionError(message)
​​​
arguments: null
​​​
caller: null
​​​
length: 1
​​​
name: ""
​​​
prototype: Object { constructor: ConnectionError(message), stack: "" }
​​​
<prototype>: function ()
​​
stack: ""
​​
<prototype>: Error.prototype
​​​
constructor: function Error()
​​​
message: ""
​​​
name: "Error"
​​​
stack: ""
​​​
toString: function toString()
​​​
<get stack()>: function stack()
​​​
<set stack()>: function stack()
​​​
<prototype>: Object { … }

[react-ga] ReactGA.initialize must be called first or GoogleAnalytics should be loaded manually warn.js:2

Uncaught (in promise) 
{…}
​
closeEvent: close
​​
bubbles: false
​​
cancelBubble: false
​​
cancelable: false
​​
code: 1005
​​
composed: false
​​
currentTarget: null
​​
defaultPrevented: false
​​
eventPhase: 0
​​
explicitOriginalTarget: WebSocket
​​​
binaryType: "blob"
​​​
bufferedAmount: 0
​​​
extensions: ""
​​​
onclose: ()
​​​​
length: 1
​​​​
name: "bound "
​​​​
<prototype>: function ()
​​​
onerror: ()
​​​​
length: 1
​​​​
name: "bound "
​​​​
<prototype>: function ()
​​​
onmessage: function ()
​​​
onopen: function ()
​​​
protocol: ""
​​​
readyState: 3
​​​
url: "ws://pirateaudio.local:6680/mopidy/ws/"
​​​
<prototype>: WebSocketPrototype { close: close(), send: send(), url: Getter, … }
​​
isTrusted: true
​​
originalTarget: WebSocket { url: "ws://pirateaudio.local:6680/mopidy/ws/", readyState: 3, bufferedAmount: 0, … }
​​
reason: ""
​​
returnValue: true
​​
srcElement: WebSocket { url: "ws://pirateaudio.local:6680/mopidy/ws/", readyState: 3, bufferedAmount: 0, … }
​​
target: WebSocket { url: "ws://pirateaudio.local:6680/mopidy/ws/", readyState: 3, bufferedAmount: 0, … }
​​
timeStamp: 568533
​​
type: "close"
​​
wasClean: true
​​
<get isTrusted()>: isTrusted()
​​​
length: 0
​​​
name: "get isTrusted"
​​​
<prototype>: function ()
​​
<prototype>: CloseEventPrototype
​​​
code: 
​​​
constructor: ()
​​​​
length: 1
​​​​
name: "CloseEvent"
​​​​
prototype: CloseEventPrototype { wasClean: Getter, code: Getter, reason: Getter, … }
​​​​
Symbol(Symbol.hasInstance): function Symbol.hasInstance()
​​​​
<prototype>: function ()
​​​
reason: 
​​​
wasClean: 
​​​
Symbol(Symbol.toStringTag): "CloseEvent"
​​​
<get code()>: function code()
​​​
<get reason()>: function reason()
​​​
<get wasClean()>: function wasClean()
​​​
<prototype>: EventPrototype { composedPath: composedPath(), stopPropagation: stopPropagation(), stopImmediatePropagation: stopImmediatePropagation(), … }
​
message: "WebSocket closed"
​
name: "ConnectionError"
​
stack: ""
​
<prototype>: {…}
​​
constructor: ConnectionError(message)
​​​
arguments: null
​​​
caller: null
​​​
length: 1
​​​
name: ""
​​​
prototype: {…}
​​​​
constructor: function ConnectionError(message)
​​​​
stack: ""
​​​​
<prototype>: Error.prototype
​​​​​
constructor: function Error()
​​​​​
message: ""
​​​​​
name: "Error"
​​​​​
stack: ""
​​​​​
toString: function toString()
​​​​​
<get stack()>: function stack()
​​​​​
<set stack()>: function stack()
​​​​​
<prototype>: Object { … }
​​​
<prototype>: ()
​​​​
apply: function apply()
​​​​
arguments: 
​​​​
bind: function bind()
​​​​
call: function call()
​​​​
caller: 
​​​​
constructor: function Function()
​​​​
length: 0
​​​​
name: ""
​​​​
toString: function toString()
​​​​
Symbol(Symbol.hasInstance): function Symbol.hasInstance()
​​​​
<get arguments()>: function arguments()
​​​​
<set arguments()>: function arguments()
​​​​
<get caller()>: function caller()
​​​​
<set caller()>: caller()
​​​​​
length: 1
​​​​​
name: "set caller"
​​​​​
<prototype>: function ()
​​​​
<prototype>: Object { … }
​​
stack: ""
​​
<prototype>: Error.prototype { stack: "", … }


```

**Screenshots**
Screenshot after reloading the Playlists section to check if any tracks were added.

<img width="1234" alt="Screen Shot 2020-11-23 at 6 52 07 PM" src="https://user-images.githubusercontent.com/10551112/100044401-7a157a80-2dc4-11eb-98a3-b0d3f1ed0e07.png">

**Environment details:**
 - OS: macOS 11.0.1 (20B29)
 - Browser Firefox 83.0 (64-bit), Safari Version 14.0.1 (16610.2.11.51.8)
 - Iris version 3.54.2

**Comment by mattallard** [comment]
Was just about to add my own issue for this, but this seems to cover the issue I am seeing as well.

Running on Raspberry Pi have seen this in the mopidy error log, not sure if useful ...

`Nov 24 07:48:03 pi-001 mopidy[19934]: DEBUG    [HttpServer] mopidy.http.handlers Received WebSocket message from 192.168.1.101: '{"method":"core.playlists.save","params":{"playlist":{"__model__":"Playlist","uri":"m3u:Christmas%20Playlist.m3u","name":"Christmas Playlist","tracks":[{"__model__":"Track","uri":"file:///mnt/usb/share/Music/Compilations/Christmas%20Songs%20for%20the%20Car/01%20Driving%20Home%20for%20Christmas%20%282009%20Remaster%29.m4a","name":"01 Driving Home for Christmas (2009 Remaster)"},{"__model__":"Track","uri":"local:track:Compilations/Legends_%20The%20Christmas%20Collection/38%20%27Twas%20the%20Night%20Before%20Christmas.m4a","sort_id":0}],"last_modified":1606064891842}},"jsonrpc":"2.0","id":13}'`
`Nov 24 07:48:03 pi-001 mopidy[19934]: ERROR    [HttpServer] mopidy.http.handlers WebSocket request error: __init__() got an unexpected keyword argument 'sort_id'`
`Nov 24 07:48:03 pi-001 mopidy[19934]: DEBUG    [HttpServer] mopidy.http.handlers Closed WebSocket connection from 192.168.1.101
N`

**Comment by jplegat** [comment]
Just to add some background here, I'm running it as part of a [Pirate Audio](https://learn.pimoroni.com/tutorial/sandyj/getting-started-with-pirate-audio) installation. As a workaround, I've rolled Mopidy-Iris - `sudo pip3 uninstall Mopidy-Iris ; sudo pip3 install Mopidy-Iris==3.53.1` - back to [3.53.1](https://github.com/jaedb/Iris/tree/3.53.1) and creating playlists and adding tracks to them is work again.

**Comment by robcee** [comment]
having a similar issue with playlists not even showing up in Iris. Selecting "Playlists" from the left navigation tree brings up a black page with no content.

**Comment by Curtis-Fletcher** [comment]
I'm getting this (as a new Mopidy/Iris user). Looking into the code it seem as though "sort_id" isn't a valid key in the mopidy "Track" model, either it's supposed to be Iris-webside internal and not be sent as part of the "core.playlist.save" method call or the Iris dev knows something about a future mopidy version that isn't in the latest build nor the HEAD of the mopidy github repo. I'm currently trying to see if there is any easy way to just strip it from the json

**Comment by jaedb** [comment]
@robcee are you able to create a new issue report for that? I suspect your experience is unrelated to this issue (which occurs when a track is added to a playlist, rather than navigating to the _Playlists_ menu).

@Curtis-Fletcher the issue is with a Mopidy request receiving a `sort_id` parameter that it isn't expecting. You're right though, `sort_id` is an Iris-exclusive track property which shouldn't be getting sent to Mopidy. There is a fix in the pipeline (67df754efec368d24eff3a550c1469ac2b09880d) to address this.

**Comment by Curtis-Fletcher** [comment]
I think I've found the problem in mopidy/middleware.js:1197 the case for "MOPIDY_ADD_PLAYLIST_TRACKS" runs injectSortId then merges the tracklist with the one recovered from mopidy, but then doesn't strip the sort_id before commiting the new list. Not sure if the stripping should be done there though, seems like there should be a generic mechanism for stripping iris-only data from model structures if they are going to be passed to "request" verbatim.

**Comment by Curtis-Fletcher** [comment]
@jaedb I just hand edited the linked fix into my webpacked app.js and I'm not sure it does everything needed, it does indeed stop the fetched tracks from getting sort_id injected but it doesn't filter the already existing playlist.tracks that are merged at line 1191 in the newese middleware.js and it seems like any new tracks that come from Iris will already have a sort_id and that will be passed through. I think another map is needed on the playlist.tracks (Though pythons map syntax is not totally clear to me as python is not my forte)

**Comment by Curtis-Fletcher** [comment]
Though, second thought. I think that would strip the sort_id from Iris' internal representation as well. Maybe you _do_ injectSortId into action.tracks_uris and strip the "__model__" part them merge with playlist.tracks and reassign back to playlist.tracks then create a new array from a map of playlist.tracks keeping only the uri's and injecting the "___model___" partand sent that new array

**Comment by Curtis-Fletcher** [comment]
I hand-edited the fix into my app.js (and switched on test mode to not use the minified one). This fixes the problem and I can add tracks to playlists now.

**Comment by robcee** [comment]
> @robcee are you able to create a new issue report for that? I suspect your experience is unrelated to this issue (which occurs when a track is added to a playlist, rather than navigating to the _Playlists_ menu).

I think so, though I'm not entirely sure how I got into this state. I will take a crack at it.

**Comment by jaedb** [comment]
Resolved in `3.55.0`
